File: 028-str-str/Solution.java
class Solution28 {
    public int strStr(String haystack, String needle) {
        //宣告雙指針以及錨點
        int i = 0;
        int j = 0;
        int Anchor = 0;
        while (i<haystack.length()){
            System.out.println(i + " " + j + " " + Anchor);
            //如果不匹配
            if (haystack.charAt(i) != needle.charAt(j)){
                System.out.println("不匹配 "+ i +" "+ j +" "+ Anchor);
                //雙指針重置
                Anchor += 1;
                j = 0;
                i = Anchor;
            }
            //如果匹配
            else {
                //如果全部匹配
                if (j == needle.length()-1){
                    return Anchor;
                }
                //雙指針往前推進
                i += 1;
                j += 1;
            }

        }
        //沒找到匹配的下標
        return -1;
    }

    static void main(String[] args){
        String s1 = ("Mississippi");
        String s2 = ("issip");
        Solution28 sol = new Solution28();
        System.out.println(sol.strStr(s1,s2));
    }
}

File: 028-str-str/solution.py
#雙指針 + while迴圈
class Solution1122:
    def strStr(self, haystack: str, needle: str) -> int:

        #宣告雙指針
        j = 0
        i = 0
        #宣告一個標籤
        tag = 0

        while(i < len(haystack)):
            print(i, j, tag)
            #如果不匹配
            if haystack[i] != needle[j]:

                #把標籤貼到下一組
                tag += 1
                #重置雙指針
                j = 0
                i = tag
                print("不匹配", i, j, tag)


            #如果匹配
            else:
                #如果完全匹配
                if(j == len(needle)-1):
                    #返回指針位置
                    return tag
                #雙指針往前推進
                i += 1
                j += 1

        #沒找到匹配的下標
        return -1

result = Solution1122().strStr("leetcode", "leeto")
print(result)